[rfxcom] Added Livolo scene support, “Livolo On/Off 1-10” subtype in Lighting 5

ashway83 opened the pull request:
Added Livolo scene support.
It's possible to trigger 4 scenes and also using Scene Switch channel we can mimic a real switch.
The difference between standard (toggle) button code (Unit 1-10 in terms of RFXCOM) and scene button code is that scene code is associated with the switch state in the moment of synchronization/learning. If switch is in On state - the scene code will be triggering On command and vice versa - if the switch is in Off state when learning the code, it will work as Off code.

Jamstah commented on addons/binding/org.openhab.binding.rfxcom/ESH-INF/thing/channels.xml:
```diff
@@ -170,4 +170,24 @@
+		<description>Scene channel</description>
+		<state readOnly="false">
+			<options>
+				<option value="0">Scene 1</option>
```
These options look odd to me, where do they come from?

ashway83 replied:
Let's go into details then.
This PR implements the following:
1. Added “Livolo On/Off 1-10” subtype in Lighting 5 which corresponds to the function group in RFXmngr.
![image](https://cloud.githubusercontent.com/assets/16033193/21762373/0006878e-d661-11e6-978f-b56d76138655.png)
2. Within the new subtype scene support was implemented:
![image](https://cloud.githubusercontent.com/assets/16033193/21762464/6a6fa2fe-d661-11e6-841c-2b2d23b4f1e5.png)
    a. Scene channel accepts numeric value 0..3 to trigger corresponding scene. Options were added for convenience, option naming is aligned with the RFXCOM Manager application. 
    b. Scene Switch functionality was implemented to address one of the Livolo protocol limitations.
The problem is that for Livolo, ON command:
`    public enum Commands {
        OFF(0x00),
        ON(0x01),`
is in fact a toggle command, meaning that if switch is in ON state, this command will turn it off. While OFF command is "all OFF" command which turns off all switches synchronized to the same (remote control) device ID. That's being said, there is no pure ON/OFF commands for Livolo.
Since scene code is associated with the switch state in the moment of synchronization/learning, we can program e.g. Scene 1 to turn on the switch and Scene 2 to turn it off accordingly. This way our Scene Switch will be acting as a normal switch similar to existing Command channel with the difference of correct behavior regardless of the switch state.

Jamstah replied:
a. Great, I guess we should stick to those rather than trying to put in something more sensible.

b. So is the idea here that if the user adds the scene switch to their dashboard, they can set it up so that scene1 is off and scene2 is on (or similar), so that they can have a reliable on/off rather than a toggle? You seem to be adding an abstraction layer on top of the scene commands, which could also be achieved using an abstraction layer in the rules (haven't tested the rules):

Switch SceneSwitch

rule "SceneOn"
when
Item SceneSwitch received command ON
then
livolo.send(SCENE1)
end

rule "SceneOff"
when
Item SceneSwitch received command OFF
then
livolo.send(SCENE2)
end

Would that achieve the same goal, or can we do more by adding it to the code? If the java version doesn't add anything above the rules, then I would recommend taking it out of the java for ease of maintenance.

ashway83 replied:
b. Yes, your understanding is correct. Although, the same behavior could be achieved using rules, I've intentionally implemented scene switch abstraction layer for the following reasons:
1) Existing Command channel doesn't provide reliable on/off functionality as mentioned above and Scene Switch is intended to address that limitation.
2) It will be pretty straightforward for the users to "reconnect" switches on their dashboards to the different channels (Scene Switch instead of Command) to start using new reliable ON/OFF functionality rather than creating extra rules. It's just a matter of few clicks in PaperUI. We need to keep in mind that as IoT is getting to the masses, there will be more and more users with limited skills in scripting/development that prefer GUI way of doing things. And that's one of the goals of openHAB 2 from my understanding.

ashway83 replied:
Regarding the rules, just tested these:

```
Switch SceneSwitch
Number LivoloScene

rule "SceneOn"
when
	Item SceneSwitch received command ON
then
	LivoloScene.send(new DecimalType(0))
end

rule "SceneOff"
when
	Item SceneSwitch received command OFF
then
	LivoloScene.send(new DecimalType(1))
end
```

Scene channel's item type is Number, so we have to pass a DecimalType. I've intentionally designed it this way, as use of string values in interfaces is usually considered to be a bad practice.

Although rules allow to mimic a scene switch, I insist to keep it on Java side for the reasons mentioned above.

Jamstah replied:
I would prefer to keep the abstraction out of the binding itself. I feel like the binding is purely for mapping the RFXCOM to items, adding abstraction to that would make the code more complicated than it needs to be. The abstraction is possible, and more flexible, with rules. @kaikreuzer this might be more of an architectural question. How do you feel about putting this kind of functionality into the binding rather than relying on rules?

I do get your point about there not being an ON, but using a scene to replace ON only works if the scene is correctly configured anyway, so it is potentially more confusing to users that way. Maybe what we should do is add a TOGGLE command to the list, TOGGLE(0x01, LIVOLO_APPLIANCE), so that it is clear than the LIVOLO has no "ON" command.

ashway83 replied:
I do like your idea of adding TOGGLE command.
Regarding adding an abstraction of scene switches, we already have a layer of abstraction to some extent in the binding. E.g. there is a command OnOff channel, while in RFXCOM command is a numeric value (byte), etc. I agree this is more architectural question, so let's see what @kaikreuzer says on this.

In the meantime I'll commit tests and sign-off my commits as required.

kaikreuzer replied:
I am not sure if I fully understood the problem. But if the device only supports toggle and there is no way to tell in which state it is, I would also go for implementing a toggle channel, although this isn't ideal either.
From the architectural view, it is no issue to add "software functionality" into a binding, i.e. abstracting some functionality that the raw hardware does not provide. If this only works when the user knows that he has to set up his device in a certain way (i.e. configuring his scenes in RFXCOM in a special way), this is indeed a bit nasty and should rather be avoided.

Jamstah commented:
It's not really something that the hardware doesn't support that we're filling in for. It is named in a way that isn't really confusing - SceneSwitch to me does convey that it is a switch between two scenes, although I suppose it could also be read as turning one scene on and off as well. It would need to be documented, but then everything does.

My only concern is that I would consider the functionality to be the role of rules, and having it two places adds maintainance overhead.

Device supports toggle and off, but not on. Maybe it should be for switching between one scene on and off, then it would definitely be an on/off replacement to overcome the missing on command?

Jamstah commented:
The "clean coder" in me is still saying do it with rules though :)

ashway83 commented:
> Device supports toggle and off, but not on. Maybe it should be for switching between one scene on and off, then it would definitely be an on/off replacement to overcome the missing on command?

To be precise, Livolo devices support toggle, group off and scene commands. There are 4 scene activation commands per remote control (Device Id in terms of our binding), there are no such thing as scene on/off.
The scene commands should be seen as a variation of true ON/OFF commands as they could be programed as ON or OFF commands during the learning process. This allows to use Scene Switch as a true Switch. The only difference from the end-user perspective is an extra step during the learning/synchronization process, as we need to learn ON and OFF commands separately (similar way we pair toggle button).

My main concern about rules vs. having Scene Switch is that I doubt many users would come up with the idea to create rules that allow to achieve true ON/OFF functionality. While Scene Switch natively addresses one of the most annoying Livolo limitations.
I know people who are using Livolo scenes, but they have never though to use them as a way to achieve reliable ON/OFF functionality.

Jamstah commented:
So what do you think of changing from switching between scene1 and scene2, to switching from (eg) scene1 to group off? Would users then only need to program the ON and not the OFF?

ashway83 commented:
I've just checked - it doesn't work that way.
Keep in mind that for Livolo `public enum Commands { OFF(0x00)` is in fact a "all OFF" or to be precise "group OFF" command which turns off all switches in the group. All switches controlled by the same remote control belong to the same group (e.g. Device Ids: 1001.x, 1001 is a remote control id, x ranges 0..10 - toggle button id, a Unit Code in RFXCOM, see the screenshot in one of my previous posts).
So, in order for the "all OFF" command to work, the switch has to be paired with the toggle command - existing ON(0x01) command.
Scene commands are independent from the toggle pairings.

ashway83 commented:
BTW, I believe many people will be surprised if they add openHAB things with the same remote control ids, e.g. 1001.1 and 1001.2. In that case both things will be turning off all switches, while toggling only the ones that they are paired with.

Jamstah commented:
Yes, I can see how that wouldn't work then! It's amazing how different all the devices are. Further in main thread, so this history didn't get lost in a review comment.

Jamstah commented:
Oh, we are in the main thread. OK.

So if we do go with this approach, have you thought about what documentation should go alongside the pr? What updates should we make to the wiki to help users get a working setup?

ashway83 commented:
I suggest to add the description of Livolo specifics (Scene Switch vs. Command channel, more details on Toggle and All Off) to the "Specifics for PacketTypes and SubTypes (PacketType.SubType)" of the wiki.
I'll draft something.
Let me also add tests.
Add I'm going to add TOGGLE command as we've discussed before, for LIVOLO and LIVOLO_APPLIANCE subtypes to make it clear how it differs from the Scene Switch.

ashway83 commented:
Added tests and ready to update Wiki as soon as PR is merged.

Jamstah commented:
You need to add a list of subtypes to the current on and off, otherwise the wrong command will be picked up when using Commands.fromByte with livolo devices. I think this will affect the test where you parse the message as well, it should come back with toggle rather than on.

Jamstah commented:
Might be able to remove the from the full set using a collections util rather than have to list them all in the code? I'll leave it up to you.

ashway83 commented:
> You need to add a list of subtypes to the current on and off, otherwise the wrong command will be picked up when using Commands.fromByte with livolo devices.

Not really. If you check carefully, Commands.fromByte() is called from encodeMessage() method which in turn is called for input RFXCOM messages only, but Livolo devices are receive only.
For output messages subtype is not verified.

It would make sense to add something like this to convertFromState() method:
```
if (subType == SubType.LIVOLO || subType == SubType.LIVOLO_APPLIANCE) {
    command = (type == OnOffType.ON ? Commands.TOGGLE : Commands.ALL_OFF);
} else {
    command = (type == OnOffType.ON ? Commands.ON : Commands.OFF);
}
```
but it doesn't make sense to do so, as byte values of ON and TOGGLE and OFF and ALL_OFF are equal.
I decided to keep enum definitions for TOGGLE and ALL_OFF for future use.

Jamstah commented:
The devices are receive only but the remote control transmits, which is what we would be using encodeMessage() for.

Either way, we should always be able to encode/decode a message and get the same thing out that we put in.

ashway83 commented:
> The devices are receive only but the remote control transmits, which is what we would be using encodeMessage() for.

In fact, encodeMessage() is never used for Livolo. The remote control responds with the confirmation package which has different format and is not processed by RFXComLighting5Message.

On the other hand, I agree that encode/decode must be reversible.

Also, I don't like the idea of limiting scope of ON and OFF to all except LIVOLO, LIVOLO_APPLIANCE. I'm not sure these commands will work similar way for all other subtypes.

I've removed TOGGLE and ALL_OFF definitions for now as they brought more confusion rather than clarification. Will add the explanation to the Wiki instead.

The current state of PR has tests and updated README file.

Jamstah commented:
Its a shame to lose the TOGGLE and ALL_OFF, I liked that it was being more specific.

There may be some future problems with maintaining state due to having two channels whose values affect each other, but we're not doing much with sending commands based on state changes in this area yet.

For me, the code here is an improvement and doesn't break anything, and documentation updates are planned to match, so LGTM.

As one final review comment, it might be worth changing from SCENE_SWITCH2 to SCENE_SWITCH_ROOM2 to make the differences between the switches more obvious to developers, as they will match better with the SCENE names.

martinvw commented:
> Will add the explanation to the Wiki instead.
@ashway83 I believe the wiki is in addons/binding/org.openhab.binding.rfxcom/README.md

Can you also look at the conflicts and Jamstah's final remark:

> As one final review comment, it might be worth changing from SCENE_SWITCH2 to SCENE_SWITCH_ROOM2 to make the differences between the switches more obvious to developers, as they will match better with the SCENE names.

martinvw commented:
@ashway83 do you expect to wrap this up, or would you like someone else to take it over?

Something for your maybe @Jamstah 🙂 

Otherwise I will try to pick it up sometime.

Jamstah commented:
Hey @martinvw, long time no see. I wouldn't really be comfortable looking at it without a device to test against, or at least someone who would be happy to do testing with me.

ashway83 commented:
Hi guys, last time it took long time to get all the approvals, so I gave
up. I'll be looking into this functionality again in the next few weeks
(need to do some tweaks for my home) and will try to wrap up the change
request in question.
Thanks.

On Mon, Mar 4, 2019 at 6:17 PM James Hewitt <notifications@github.com>
wrote:

> Hey @martinvw <https://github.com/martinvw>, long time no see. I wouldn't
> really be comfortable looking at it without a device to test against, or at
> least someone who would be happy to do testing with me.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openhab/openhab2-addons/pull/1672#issuecomment-469313302>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/APSlqQgjX2XGacBTQ-G9iDhg_WyWawmaks5vTUcLgaJpZM4LdrUt>
> .
>

davidgraeff commented:
So this is still active if I understood correctly?

J-N-K commented:
Closing due to inactivity. Please open a new PR if you want to continue this work.
